[scroll-animations-1] Support for multiple scroll offsets #4912

ogerchikov opened the pull request:
Described specification for ScrollTimeline.scrollOffsets API.

kevers-google reviewed (commented):
Sorry for the tardy response. Overall looks good. Deferring to Rob and Anders for approval.

andruud commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,66 @@ follows:
+    1.  Set <var>first offset</var> to false.
+1.  For each <var>scroll offset</var> in the list of
+    {{ScrollTimeline/scrollOffsets}}, perform the following steps:
+    1.  Let <var>effective offset</var> be a result of applying the procedure
```
"Let X be _a_ result of applying Y" is pretty uncommon. (Occurs several times here). 

"Let X be _the_ result of applying Y"?

ogerchikov replied:
Thanks! Done.

andruud commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,66 @@ follows:
+    1.  Run the procedure to [=resolve a scroll timeline offset=] for 'auto'
+        with the <var>is first</var> flag set to <var>first offset</var> and
+        add the resulted value into <var>effective scroll offsets</var>.
+1.  If {{ScrollTimeline/scrollOffsets}} has one element
```
Should this say:

If {{ScrollTimeline/scrollOffsets}} has at least one element?

ogerchikov replied:
This describes how to handle single scroll offset - add 'auto' for as a first element, the provided offset will be added as part of the next statement.

andruud replied:
Right. I did understand that at some point, then I confused myself.

Then probably this should say:

If {{ScrollTimeline/scrollOffsets}} has _exactly_ one element

Just to make it clear that we're not intending the "at least" meaning. (Which would prepend 'auto' for any non-empty list I guess).

ogerchikov replied:
Done

andruud commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,66 @@ follows:
+    <code>(<var>offset id</var> - 1 + <var>interval progress</var>) &times;
+    <var>offset weight</var></code>.
+
+Note: The rational behind choosing last matching offset in the array is to
```
rational -> rationale

ogerchikov replied:
Done.

andruud commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,66 @@ follows:
+
+1.  Let <var>scroll offsets</var> be a result of applying the procedure to
+    [=resolve scroll timeline offsets=] for {{scrollOffsets}}.
+1.  Let <var>offset id</var> correspond to the position of the last offset in
```
Suggestion: `<var>offset index</var>`

ogerchikov replied:
Done.

andruud commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,66 @@ follows:
+    [=resolve scroll timeline offsets=] for {{scrollOffsets}}.
+1.  Let <var>offset id</var> correspond to the position of the last offset in
+    <var>scroll offsets</var> whose value is less than or equal to
+    <var>offset</var> and the value at the following position greater than
```
Does this mean that the specified scroll offsets have ordering constraints? If yes, what are they, and when/how are they enforced?

ogerchikov replied:
There is no ordering constrains. Explanation why last interval is chosen is explained in the note below the section. I added another note explaining why matching interval will always be found.

andruud commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,66 @@ follows:
+    <code>(<var>offset</var> - <var>start offset</var>) /
+    (<var>end offset</var> - <var>start offset</var>)</code>.
+1.  Return the result of evaluating
+    <code>(<var>offset id</var> - 1 + <var>interval progress</var>) &times;
```
> `<var>offset id</var> - 1`

I guess this means that `offset id` starts at `1`? Can't we just specify earlier that it starts at zero?

ogerchikov replied:
Thanks for pointing this out. It's a copy and paste from the [implementation](https://source.chromium.org/chromium/chromium/src/+/master:cc/animation/scroll_timeline.h;l=149;drc=131697ea2c25234744fe1917571bfa454c2d22cb?q=scroll_timeline&ss=chromium) where offset_id correspond to the end of the matching interval.
I added a comment to [crbug](https://bugs.chromium.org/p/chromium/issues/detail?id=1094014#c7) to update the implementation based on spec.

andruud reviewed (commented):
( One month later...: )

ogerchikov commented on scroll-animations-1/Overview.bs:
```diff
@@ -967,15 +987,14 @@ instead of document's scrolling Element for `auto`.
-
 <pre class='descdef'>
-  Name: start
+  Name: scrollOffsets
```
How about `scroll-offsets` to match Javascript definition of ScrollTimeline?

andruud replied:
Seems somewhat unnecessary to have the `scroll-` prefix when we're already inside `@scroll-timline`. IIRC we had `scroll-time-range` etc before, but the prefix was dropped.

_But_, if there are reasons for keeping the `scroll-` prefix on `scroll-offsets` specifically, then `scroll-offsets` WFM.

flackr replied:
There was a concern that offset has a different meaning in animation context, to specify the position of each keyframe. We may in fact want to add offsets as well as scrolloffsets in the future so that you could map scroll offsets to arbitrary animation offsets rather than having them uniformly distributed.

ogerchikov replied:
Changed to scroll-offsets.

andruud replied:
I see, makes sense.

ogerchikov commented:
Thanks @andruud for reviewing! I addressed your comments besides the one which refers to `scrollOffsets` name. Before I make the change I would like to get consensus from other reviewers. Please take a look at the latest patch.

flackr commented on scroll-animations-1/Overview.bs:
```diff
@@ -506,9 +484,11 @@ There are two types of scroll timeline offset: [=container-based offset=], and
-resolved from {{start}}. The <dfn>effective end offset</dfn> is
-the [=effective scroll offset=] that is resolved from {{end}}.
+The <dfn>effective start offset</dfn> is the value of first offset in
+[=effective scroll offsets=] array or null if the array is empty.
```
What if the array has a single value in it? Shouldn't the effective offset be 'auto'?

flackr replied:
Nevermind, on rereading the algorithm for calculating effective scroll offsets, it looks like the one element offsets case becomes two effective scroll offsets.

flackr commented on scroll-animations-1/Overview.bs:
```diff
@@ -506,9 +484,11 @@ There are two types of scroll timeline offset: [=container-based offset=], and
+[=effective scroll offsets=] array or null if the array is empty.
+
+The <dfn>effective end offset</dfn> is the value of last offset in
+[=effective scroll offsets=] array or null if the array is empty.
```
Should an empty array produce effective offsets of [auto, auto]?

flackr commented on scroll-animations-1/Overview.bs:
```diff
@@ -789,28 +770,72 @@ follows:
+calculation.
+
+Note: The rationale behind choosing last matching offset in the array is to
+be consistent with CSS property overrides.
```
I think we should link to the rule for finding which KeyframeRule applies in a comma separated list of keyframes with overlap as justification: https://drafts.csswg.org/css-animations-1/#interface-csskeyframesrule-findrule

ogerchikov replied:
Done.

ogerchikov commented on scroll-animations-1/Overview.bs:
```diff
@@ -294,8 +288,7 @@ if (window.matchMedia('(prefers-reduced-motion: no-preference)').matches) {
     orientation: "vertical",
-    scrollOffset: '0px',
-    end: '200px'
+    scrollOffsets: '0px', '200px'
```
Done.

ogerchikov commented on scroll-animations-1/Overview.bs:
```diff
@@ -156,15 +156,13 @@ Using the CSS markup:
     orientation: block;
-    start:  200px;
-    end: 300px;
+    scrollOffsets:  200px, 300px;
```
Done.

ogerchikov commented:
Thanks for reviewing! All your comments are addressed in the recent commit.

birtles reviewed (approved):
If you've got review from @kevers-google, @andruud, and @flackr then I'm pretty sure you don't need mine! (I did skim it however, and it looked very thorough.)

foolip commented:
Was there a Chromium bug filed for this? `CSSScrollTimelineRule` was changed to remove `start` and `end` properties, but those are still in Chromium.

ogerchikov commented:
@foolip Yes, https://bugs.chromium.org/p/chromium/issues/detail?id=1094014 includes removal of start and end properties.

foolip commented:
Thanks @ogerchikov! I noticed the removal when updating IDL for https://mdn-bcd-collector.appspot.com/, and wasn't sure if I should add back some custom tests for this or not. I won't then, if it's not going to be left long in Chromium.
